src/operator/numpy/np_trace_op-inl.h

```diff
@@ -0,0 +1,255 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file np_trace_op-inl.h
+ * \brief Function definition of matrix numpy-compatible trace operator
+ */
+
+#ifndef MXNET_OPERATOR_NUMPY_NP_TRACE_OP_INL_H_
+#define MXNET_OPERATOR_NUMPY_NP_TRACE_OP_INL_H_
+
+#include <dmlc/parameter.h>
+#include <mxnet/operator_util.h>
+#include <vector>
+#include <utility>
+#include <algorithm>
+#include "../mxnet_op.h"
+#include "../operator_common.h"
+#include "../elemwise_op_common.h"
+#include "../tensor/broadcast_reduce_op.h"
+
+namespace mxnet {
+namespace op {
+
+struct NumpyTraceParam: public dmlc::Parameter<NumpyTraceParam> {
+    int offset, axis1, axis2;
+    DMLC_DECLARE_PARAMETER(NumpyTraceParam) {
+        DMLC_DECLARE_FIELD(offset)
+            .set_default(0)
+            .describe("Offset of the diagonal from the main diagonal. "
+                      "Can be both positive and negative. Defaults to 0.");
+        DMLC_DECLARE_FIELD(axis1)
+            .set_default(0)
+            .describe("Axes to be used as the first axis of the 2-D sub-arrays "
+                      "from which the diagonals should be taken. Defaults to 0.");
+        DMLC_DECLARE_FIELD(axis2)
+            .set_default(1)
+            .describe("Axes to be used as the second axis of the 2-D sub-arrays "
+                      "from which the diagonals should be taken. Defaults to 1.");
+    }
+};
+
+template<int ndim, int req, bool back>
+struct numpy_trace {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, DType* out, const DType* a,
+                                  mshadow::Shape<ndim> oshape,
+                                  mshadow::Shape<ndim> ishape,
+                                  index_t stride, index_t offset, int dlength) {
+    using namespace mxnet_op;
+    using namespace mshadow;
+    index_t j = ravel(unravel(i, oshape), ishape) + offset;
+    if (back) {
+      for (index_t k = 0; k < dlength; ++k) {
+        KERNEL_ASSIGN(out[j], req, a[i]);
+        j += stride;
+      }
+    } else {
+      if (req == kWriteTo) {
+         out[i] = 0;
+        for (index_t k = 0; k < dlength; ++k) {
+          out[i] += a[j];
+          j += stride;
+        }
+      } else if (req == kAddTo) {
+        for (index_t k = 0; k < dlength; ++k) {
+          out[i] += a[j];
+          j += stride;
+        }
+      }
+    }
+  }
+};
+
+
+template<typename xpu, bool back>
+void NumpyTraceOpProcess(const TBlob& in_data,
+                   const TBlob& out_data,
```

Comment:
please align this line with the argument in the line above. like:
```c++
void NumpyTraceOpProcess(const TBlob& in_data,
                         const TBlob& out_data,
```

Reply:
Same for all arguments below and for all other function signatures.

Reply:
fixed